src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,104 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size;
+    private int capacity;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+        capacity = DEFAULT_CAPACITY;
+    }
+
     @Override
-    public void add(T value) {
+    protected Object clone() throws CloneNotSupportedException {
+        T[] clonedArray = (T[]) super.clone();
+        System.arraycopy(elements, 0, clonedArray, 0, size);
+        return clonedArray;
+    }
 
+    @Override
+    public void add(T value) {
+        resize();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of range: " + index);
+        }
+        resize();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        resize();
```

Comment:
```suggestion
```
not fixed
resizing happens in add() method anyway
it's functionally equivalent to calling add multiple times